### PR TITLE
Fine-tune Decentralized communications

### PR DESCRIPTION
**Pull request by K1li4nL**
Two things in this pr: 

`GetMessagesByIdResponseHanlder`:
The validation has changed so that instead of retrying a certain amount of time per channel, each message has its own counter. 

`ConnectionMediator`: 
Prevent it to connect to the same urls upon change in the config. The behavior is not perfect as for now there is no link between clientActorRef and their urls. `GreetServer` will be helpful in that regard.

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 1]
```diff
@@ -43,65 +44,60 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
     }
-
-    failedGraphMessages
   }
 
-  private def wrapMsgInPublish(map: Map[Channel, Set[Message]]): Map[Channel, Set[GraphMessage]] = {
-    map.map { case (channel, set) =>
-      channel -> set.map(msg =>
-        Right(JsonRpcRequest(
+  // Must return a list sorted by channel length
```
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:88]: Try to be more precise in your comments: a list of what?

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 2]
```diff
@@ -12,23 +12,24 @@ import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage, PipelineError, pretty
...
+import scala.util.Success
 import scala.concurrent.Await
 
 //This object's job is to handle responses it receives from other servers after sending a heartbeat.
 // When receiving the missing messages, the server's job is to write them on the database.
 object GetMessagesByIdResponseHandler extends AskPatternConstants {
 
-  private final val MAX_RETRY = 5
-  private final val SUCCESS = 0
+  private val MAX_RETRY_PER_MESSAGE = 5
```
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:22]: Why did you remove the `final`? (in line 22 too)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 3]
```diff
@@ -43,65 +44,60 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
+              if (retry == 1) {
+                println("Errors: " + err.toString + "\n On:\n" + prettyPrinter(graphMessage))
+              }
+            case _ => /* DO NOTHING */
           }
         }
     }
 
-    passThroughPipeline(failedMessages, validatorFlow, remainingAttempts - 1)
+    passThroughPipeline(failedMessages.reverse, validatorFlow)
```
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:73]: Why do you need to reverse your list of messages ?

**Review by Ajkunas — CHANGES_REQUESTED** [review]
Most of the comments are nitpicks or questions. Othw all good :)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 2 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:22]: I figured the `final` didn't add anything given that `val` is used and these are private variables.

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 3 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:73]: I unfortunately don't have any data to back this up but:
Since we are ordering the channels based on the length, I  wanted to keep their order. (since the we add the `graphMessage` on the top of the failed list, the order is reversed) 

I don't know performance wise if one way is better than the other though...

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 1 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala]: I added proper documentation

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 4]
```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
+  /** Wrap messages in Publish RPCs along with a counter
+    *
+    * @param map
+    *   A map of channels to their set of messages
+    * @return
+    *   A list of GraphMessage - Int pairs where the GraphMessage are sorted in increasing order of their channel length and the Int is the number of retry left for each message
+    */
+  private def wrapMsgInPublish(map: Map[Channel, Set[Message]]): List[(GraphMessage, Int)] = {
+    var publishedList = map.foldLeft(List.empty[JsonRpcRequest])((acc, elem) =>
+      acc ++ elem._2.map(msg =>
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:124]: I think you could use a `flatMap` instead of `foldLeft` + an accumulator

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 5]
```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
           MethodType.PUBLISH,
-          new Publish(channel, msg),
+          new Publish(elem._1, msg),
           Some(0)
-        ))
+        )
       )
-    }
+    )
+    publishedList = publishedList.sortWith((e1, e2) => e1.getParamsChannel.channel.length <= e2.getParamsChannel.channel.length)
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:133]: Is this ordering part of the decentralized specification? I find an ordering by the channel name's length quite strange... Is there some specific intent behind that?

Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:133]: On a scala side, you could avoir using `var` for `publishedList` by simply returning
```scala
publishedList.sortWith(...)
.map(...)
```

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 3 above)
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:90]: Is there really a need to keep that ordering all the way through the simulation pipeline?

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 6]
```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
-          // only log a during last attempt
-          if (remainingAttempts == 1) {
-            println("Errors: " + failedMsgErrorMap.map {
-              case (msg, error) => error.toString + "\n On:\n" + prettyPrinter(msg)
-            })
+      case (graphMessage, retry) =>
+        if (retry > 0) {
+          messagesThroughPipeline(validatorFlow, graphMessage) match {
+            case Left(err) =>
+              failedMessages = (graphMessage, retry - 1) :: failedMessages
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:80]: Nitpicks: You could probably use the operator `::=` that you made me discover a couple of lines above :)
On a side note, maybe using a mutable collection from `scala.collection.mutable` would be better than using a `var` for performance purposes

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 7]
```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
...
   @tailrec
   private def passThroughPipeline(
-      receivedResponse: Map[Channel, Set[GraphMessage]],
-      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed],
-      remainingAttempts: Int
+      receivedResponse: List[(GraphMessage, Int)],
+      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed]
   )(implicit system: ActorSystem): Boolean = {
-    if (receivedResponse.isEmpty || remainingAttempts <= 0) {
+    if (receivedResponse.isEmpty || receivedResponse.forall(elem => elem._2 <= 0)) {
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:70]: The second check `receivedResponse.forall(elem => elem._2 <= 0)` is actually not really necesssary as messages with `retry == 0` are ignored in the following `foreach(...)`, so if all the responses received have 0 retry left, the next `passThroughPipeline` will have an empty list of responses and will therefor still return (so it would avoid iterating over the entire list of responses 2 times per call instead of only 1)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala [hunk 8]
```diff
@@ -12,23 +12,32 @@ import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage, PipelineError, pretty
...
+    *   Implicit actor system to use the given validator
+    * @return
+    *   Left if some messages couldn't be validated after MAX_RETRY_PER_MESSAGE times, Right for success
+    */
   def responseHandler(messageRegistry: MessageRegistry)(implicit system: ActorSystem): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
     case Right(JsonRpcResponse(_, Some(resultObject), None, _)) =>
       resultObject.resultMap match {
         case Some(resultMap) =>
-          val receivedResponse: Map[Channel, Set[GraphMessage]] = wrapMsgInPublish(resultMap)
+          val receivedResponse = wrapMsgInPublish(resultMap)
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:38]: nitpick: maybe you could pass the number of retries as a parameter to `wrapMsgInPublish` as you used to do with `passThroughPipeline`: otherwise, I feel like it kinds of come from nowhere as `wrapMsgInPublish` as a name doesn't really carry the intent of allocating a number of retry to each message...

**Review by Jharaxus — DISMISSED** [review]
Nice code, it indeed makes more sense to have a retry per message !

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 5 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:133]: Absoluetely not in the protocol. The intent is as follows: 
In general shorter channel name are less specific and so should be treated first. 

`/root`: should be treated first as it contains the lao creation messages
`/root/lao_id`: should be treated right after since it contains the rollcalls and all.

It might not be very elegant and there are exceptions but the retries seem to be enough to deal with them

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala (hunk 3 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala:90]: I removed since its utility beside being confusing is questionable.

**Review by Ajkunas — APPROVED** [review]
LGTM
